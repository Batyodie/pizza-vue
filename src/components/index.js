import Button from "./Button/Button.vue";
import Card from "./Card/Card.vue";
import DropDown from "./DropDown/DropDown.vue";
import Logo from "./Logo/Logo.vue";
import ShopCart from "./ShopCart/ShopCart.vue";

export { Button, Card, DropDown, Logo, ShopCart };
