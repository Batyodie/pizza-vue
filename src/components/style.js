import BtnTag from "@/components/Button/Button.css";
import Button from "@/components/Button/Button.css";
import DropDownStyle from "@/components/DropDown/DropDown.css";
import CardStyle from "@/components/Card/Card.css";
import LogoStyle from "@/components/Logo/Logo.css";
import ShopCartStyle from "@/components/ShopCart/ShopCart.css";
import BtnStyle from "@/components/Button/Button.css";

export {
  Button,
  BtnTag,
  DropDownStyle,
  CardStyle,
  LogoStyle,
  ShopCartStyle,
  BtnStyle
};
