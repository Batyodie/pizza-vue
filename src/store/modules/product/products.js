import { getProducts } from "@/api/products.api";
import {
  SET_DISCHARGE,
  SET_LOADED,
  SET_PRODUCTS,
  SET_PRODUCTS_ERROR
} from "../../mutation-types";
export default {
  state: {
    loadedProducts: false,
    products: [],
    productError: null
  },
  mutations: {
    [SET_PRODUCTS](state, products) {
      state.products = products;
    },
    [SET_LOADED](state) {
      state.loadedProducts = true;
    },
    [SET_DISCHARGE](state) {
      state.loadedProducts = false;
    },
    [SET_PRODUCTS_ERROR](state, error) {
      state.productError = error;
    }
  },
  actions: {
    async fetchProducts({ commit, state, rootState }) {
      commit("SET_DISCHARGE");
      try {
        // get request to the api using the exported function getProducts,
        // into which we pass the id of the active pizza category and the id of the active pizza filter
        const tagCategoryID = rootState.tags.isActiveTag;
        const dropDownActive = rootState.dropdown.DropDownItemIsActive;
        const products = await getProducts(tagCategoryID, dropDownActive);

        commit("SET_PRODUCTS", products);
        if (state.products.length) {
          commit("SET_LOADED");
        } else {
          commit("SET_DISCHARGE");
        }
      } catch (err) {
        commit("SET_PRODUCTS_ERROR", err);
      }
    }
  },
  getters: {
    getCards: ({ products }) => products,
    getPizzasLoadedFlag: ({ loadedProducts }) => loadedProducts
  }
};
