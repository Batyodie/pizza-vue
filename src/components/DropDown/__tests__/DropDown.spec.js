import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import dropDown from "@/store/modules/dropdown/dropdown";
import product from "@/store/modules/product/products";
jest.mock("@/store/modules/product/products");
import { DropDown } from "@/components";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Unit tests for DropDown component", () => {
  // fake data
  // const DropDownIsOpen = false;
  const DropDownItems = [
    { id: 5, text: "популярности", type: "rating", order: "desc" },
    { id: 6, text: "по цене", type: "price", order: "desc" },
    { id: 7, text: "по алфавиту", type: "name", order: "asc" }
  ];
  const DropDownItemIsActive = {
    text: "популярности",
    type: "rating",
    order: "desc"
  };
  let store;
  let actions;

  beforeEach(() => {
    actions = {
      fetchProducts: product.actions.fetchProducts,
      DropDownOpened: dropDown.actions.DropDownOpened,
      DropDownGlobalClosed: dropDown.actions.DropDownGlobalClosed,
      DropDownSelected: dropDown.actions.DropDownSelected
    };
    store = new Vuex.Store({
      state: dropDown.state,
      actions,
      mutations: dropDown.mutations,
      getters: dropDown.getters
    });
  });

  // create wrapper factory
  const wrapperFactory = (mouning = shallowMount, { propsData } = {}) => {
    return mouning(DropDown, {
      localVue,
      store,
      propsData: {
        ...propsData
      }
    });
  };

  it("should initialize correctly dropdown component", () => {
    const wrapper = wrapperFactory();
    const icon = "/img/arrow-top.svg";

    wrapper.setData({ icon });
    // correctly Vue instance
    expect(wrapper.exists()).toBeTruthy();
    // We expect the wrapper to be the dropdown component
    expect(wrapper.is(DropDown)).toBeTruthy();
    expect(wrapper.vm.icon).toEqual(icon);
    // expect(wrapper.find(".IconImage").attributes("src")).toBe(icon);
  });

  it("should render correctly dropdown component", () => {
    const wrapper = wrapperFactory();
    const activeLable = DropDownItemIsActive.text;
    const renderLabel = `Сортировка по: 
        ${activeLable}`;

    const label = wrapper.find(".Btn");

    // expecting label wrapper to exist
    expect(label.exists()).toBeTruthy();
    // expect the label to have a activeLable
    expect(label.text(activeLable)).toBeTruthy();
    // expect the label selector to have activeLable
    expect(label.text()).toContain(activeLable);
    expect(wrapper.text()).toEqual(renderLabel);
  });

  it("should match the snapshot DropDown component", () => {
    const wrapper = wrapperFactory();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should correctly opened dropdown component", async () => {
    const wrapper = wrapperFactory(mount);
    const btn = wrapper.find(".Btn");
    // click label button
    await btn.trigger("click");

    const list = wrapper.find(".List");
    // expect the list is exists
    expect(list.exists()).toBe(true);
    // expect the list is visible
    expect(list.isVisible()).toBe(true);
  });

  it("should correctly render first opened dropdown to popup list items", () => {
    const wrapper = wrapperFactory(mount);
    const btn = wrapper.find(".Btn");
    // click label button
    btn.trigger("click");
    const items = wrapper.findAll(".Item");
    // expect the items at index to equal for DropDownItems
    expect(items.at(0).text()).toEqual(DropDownItems[0].text);
    // expect the first item to have active class
    expect(items.at(0).classes("ItemActive")).toBe(true);
    expect(items.at(1).text()).toEqual(DropDownItems[1].text);
    expect(items.at(2).text()).toEqual(DropDownItems[2].text);
  });

  it("should correctly selection dropdown item", async () => {
    const wrapper = wrapperFactory();
    const btn = wrapper.find(".Btn");
    const indexItem = 1;
    // click label button
    await btn.trigger("click");

    const items = wrapper.findAll(".Item");
    // expect the dropdown in open
    expect(store.state.DropDownIsOpen).toBe(true);
    // click to indexItem
    await items.at(indexItem).trigger("click");
    // expect the dropdown is close
    expect(store.state.DropDownIsOpen).toBe(false);
    // expect the label button text to eaual the dropdown items text
    expect(btn.text()).toEqual(DropDownItems[indexItem].text);
    // expect the actions to have be called
    expect(actions.fetchProducts).toHaveBeenCalled();
    // again click button
    await btn.trigger("click");
    // expect the indexItem to have active class
    expect(
      wrapper
        .findAll(".Item")
        .at(indexItem)
        .classes("ItemActive")
    ).toBe(true);
  });
});
