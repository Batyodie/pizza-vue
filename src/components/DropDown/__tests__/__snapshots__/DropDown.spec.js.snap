// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for DropDown component should match the snapshot DropDown component 1`] = `
<div class="DropDown">
  <div style="display: flex; align-items: center;"><img width="10" height="6" src="" alt="Arrow top" class="IconImage"> <span class="Label">Сортировка по:</span> <button class="Btn">
      популярности
    </button></div>
  <!---->
</div>
`;
